src/libraries/System.Private.CoreLib/src/System/Collections/Generic/IEnumerator.cs

```diff
@@ -16,5 +17,10 @@ public interface IEnumerator<out T> : IDisposable, IEnumerator
         {
             get;
         }
+
+        // NOTE: An implementation of an enumerator using a ref struct T will
+        // not be able to implement IEnumerator.Current to return that T (as
+        // doing so would require boxing). It should throw a NotSupportedException
+        // from that property implementation.
```

Comment:
Can we have a default implementation like:
```csharp
object IEnumerator.Current => typeof(T).IsByRefLike ? throw new NotImplementedException() : Current;
```

or would that be a breaking change?

Reply:
> Can we have a default implementation like:

We've shied away from using DIMs on very core interfaces, and when we've ventured into that territory (as we recently tried with having the mutable collection interfaces implement the readonly ones), we've had difficulties.

cc: @terrajobst, @bartonjs

Reply:
Deriving from `IEnumerable<T>` / `IEnumerable` is quite common (~18% of packages on NuGet.org are). However, my gut feeling is that it's mostly iterators (i.e. code using `yield`) rather than manually implementing the interface.

Hence, I don't think it's worth to take on the risk of a binary breaking change to make the experience for a very small group slightly better...

Reply:
Agree. I think guidance for `ref struct` authors who want to create `IEnumerable<MyAwesomeRefStruct>` is a better trade off than a DIM.

Reply:
> However, my gut feeling is that it's mostly iterators (i.e. code using yield) rather than manually implementing the interface.

What percentage of the type names are unspeakable? Shouldn't require a gut feeling :smile:

(I'm just curious; I don't think it changes the stance.)

Reply:
Sadly I can't answer that without a custom metadata collector / analysis.

Reply:
> Sadly I can't answer that without a custom metadata collector / analysis.

@marklio, is this easy for you to answer? If yes, great. If not trivial, don't worry about it.